kevm-pyk/src/kevm_pyk/utils.py

```diff
@@ -138,9 +139,59 @@ def kevm_apr_prove(
         return True
     else:
         _LOGGER.error(f'Proof failed: {_cfgid}')
+        if failure_info:
+            failure_log = print_failure_info(_cfg, _cfgid, kcfg_explore)
+            for line in failure_log:
+                _LOGGER.warning(line)
         return False
 
 
+def print_failure_info(_cfg: KCFG, _cfgid: str, kcfg_explore: KCFGExplore) -> list[str]:
+    unique_target = _cfg.get_unique_target()
+
+    res_lines: list[str] = []
+
+    num_frontier = len(_cfg.frontier)
+    num_stuck = len(_cfg.stuck)
+    res_lines.append(f'{num_frontier + num_stuck} Failure nodes. ({num_frontier} frontier and {num_stuck} stuck)')
+    if num_frontier > 0:
+        res_lines.append('')
+        res_lines.append('Frontier nodes:')
+        for node in _cfg.frontier:
+            res_lines.append('')
+            res_lines.append(f'ID: {node.id}:')
+    if num_stuck > 0:
+        res_lines.append('')
+        res_lines.append('Stuck nodes:')
+        for node in _cfg.stuck:
+            res_lines.append('')
+            res_lines.append(f'  Node id: {str(node.id)}')
+
+            simplified_node, _ = kcfg_explore.cterm_simplify(node.cterm)
+            simplified_target, _ = kcfg_explore.cterm_simplify(unique_target.cterm)
+
+            node_cterm = CTerm.from_kast(simplified_node)
+            target_cterm = CTerm.from_kast(simplified_target)
+
+            res_lines.append('  Failure reason:')
+            _, reason = kcfg_explore.implication_failure_reason(node_cterm, target_cterm)
+            res_lines += [f'    {line}' for line in reason.split('\n')]
+
+            res_lines.append('  Path condition:')
+            res_lines += [f'    {kcfg_explore.kprint.pretty_print(_cfg.path_constraints(node.id))}']
+
+            if 'foundry_success' in reason:
+                res_lines.append('')
+                res_lines.append('See `foundry_success` predicate for more information:')
+                res_lines.append(
+                    'https://github.com/runtimeverification/evm-semantics/blob/master/include/kframework/foundry.md#foundry-success-predicate'
+                )
```

Comment:
This is the only part that is specific to foundry actually in the reason here. So I would actually just remove this chunk for now, and inline this logic into the places where `print_failure_info` is called.

Then you can actually take this function and upstream it to `APRProof.failure_info` or something instead. Or perhaps as `APRProof.pretty`.

Reply:
Issue made: https://github.com/runtimeverification/pyk/issues/463

Reply:
For now I've moved the `foundry_success` outside of `print_failure_info()` and added the link to the gitbook.